Fancy Dungeon Map

kevinthegreat1 opened the pull request:
* Render player heads in the dungeon map
* Add map to the leap menu
* More to come

Note: this screenshot is taken on a low gui scale

<img width="1440" alt="Fancy Dungeon Map" src="https://github.com/user-attachments/assets/ea4e91db-e97d-46db-9e50-35ae8c9d8b5e" />

Fluboxer commented:
Can I ask for a thing? Make it show arrow to where player is looking
iirc custom fancy maps often overlook that despite this. It may help tell where players are headed and it is a thing normal vanilla map can do

AC19970 commented:
The game freezes when starting. The avatars won't be displayed if teammates are too far. Do I need to set the render distance and simulation distance higher?

kevinthegreat1 commented:
Fixed the lag spike when starting the dungeon and added a little direction indicator. The problem with players going too far is a little more complicated, and I will look into it later.

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/dungeon/secrets/DungeonPlayerManager.java:
```diff
@@ -39,6 +43,18 @@ public static DungeonClass getClassFromPlayer(String name) {
+	 */
+	private static void onDungeonLoaded() {
+		assert MinecraftClient.getInstance().world != null;
+		for (Entity entity : MinecraftClient.getInstance().world.getEntities()) {
```
This won't work because some players don't get warped into the dungeon at the start and get kicked, so this will need adjustment.

kevinthegreat1 replied:
It seems to work fine for me, and it removed noticeable lag.

AzureAaron replied:
I'm not saying it doesn't work I'm just saying it shouldn't only happen on dungeon load. Players won't warp in if they go to another island just before being warped (due to transfer cooldown) for example. You can just run the logic multiple times at different points to solve the problem.

kevinthegreat1 replied:
The problem is, if I add the original one back, where it loads after the dungeon starts, it's at the same time the map loads it. So if I add it back, it wouldn't do anything because the map would load the same thing at the same time.

kevinthegreat1 replied:
Made rendering non-blocking instead.

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/dungeon/LeapOverlay.java:
The scaling was intentional for making the leap overlay decently big, if it's too small then it is not as effective since scaling makes it easier to see and use (this was from direct feedback that my earlier iterations of it were far too small compared to implementations on 1.8). The scaling should be brought back and adjusted if necessary in order to make space for the map (see other comment).

kevinthegreat1 replied:
I added back scaling, but there's some weirdness with the positioning, and I might need some custom positioning.

kevinthegreat1 replied:
Done in 0148908.

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/dungeon/LeapOverlay.java:
```diff
@@ -99,34 +88,27 @@ public void onSlotUpdate(ScreenHandler handler, int slotId, ItemStack stack) {
+		int x = (width >> 1) - 64;
+		int y = (height >> 2) - 64;
+		DungeonMap.render(context, x, y, 1, true);
+		context.drawBorder(x, y, 128, 128, -1);
```
The map should maybe be a widget too not sure depends how this works with ensuring it remains on screen. The map appearing here shouldn't be mandatory either; it also doesn't appear to check if the player is actually in the boss or not and you wouldn't want the regular dungeon map appearing in boss.

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/dungeon/DungeonMap.java:
```diff
@@ -45,30 +53,56 @@ public static void init() {
+		context.getMatrices().pop();
     }
 
+	private static void renderPlayerHeads(MinecraftClient client, DrawContext context) {
```
Maybe it would be worth ensuring that the actual player markers don't render (since it seems they would render underneath?). Could probably extend the `MapRenderState` and add a field for this or something.

kevinthegreat1 replied:
I don't think this is worth it for now.

AzureAaron commented on src/main/java/de/hysky/skyblocker/skyblock/dungeon/LeapOverlay.java:
```diff
@@ -213,7 +178,22 @@ public void onClick(double mouseX, double mouseY) {
+
+		@Override
+		public boolean equals(Object obj) {
+			return obj instanceof PlayerReference playerRef && uuid.equals(playerRef.uuid);
```
Shouldn't the hash code also reflect this?

layou233 commented:
Got some ideas too. It would be nice if you could consider them. :)

1. An option to not render head of myself. Pretty neat and makes it easier for me to find my direction.

```diff
// src/main/java/de/hysky/skyblocker/skyblock/dungeon/DungeonMap.java
@@ -153,6 +153,7 @@
 		for (Map.Entry<String, MapDecoration> mapDecoration : ((MapStateAccessor) state).getDecorations().entrySet()) {
 			PlayerRenderState player = PlayerRenderState.of(mapDecoration);
 			if (player == null) continue;
+			if (player.uuid().equals(MinecraftClient.getInstance().player.getUuid())) continue;
 			DungeonClass dungeonClass = DungeonPlayerManager.getClassFromPlayer(player.name());
```

2. Not render the map and center the leap overlay when in boss fight. Might be useful for leaping by just move mouse to different directions and do subconscious clicking.

```diff
// src/main/java/de/hysky/skyblocker/skyblock/dungeon/LeapOverlay.java
@@ -68,7 +69,11 @@
 		}
 
 		gridWidget.refreshPositions();
-		SimplePositioningWidget.setPos(gridWidget, 0, 0, this.width, this.height, 0.5f, 0.75f);
+		float relativeY = 0.5f;
+		if (DungeonManager.isClearingDungeon()) {
+			relativeY = 0.75f;
+		}
+		SimplePositioningWidget.setPos(gridWidget, 0, 0, this.width, this.height, 0.5f, relativeY);
 		gridWidget.forEachChild(this::addDrawableChild);
 	}
 
@@ -109,10 +114,12 @@
 	public void render(DrawContext context, int mouseX, int mouseY, float delta) {
 		super.render(context, mouseX, mouseY, delta);
 
-		int x = (width >> 1) - 64;
-		int y = (height >> 2) - 64;
-		hovered = DungeonMap.render(context, x, y, 1, true, mouseX - x, mouseY - y, hoveredElement(mouseX, mouseY).filter(PlayerButton.class::isInstance).map(PlayerButton.class::cast).map(p -> p.reference.uuid()).orElse(null));
-		context.drawBorder(x, y, 128, 128, -1);
+		if (DungeonManager.isClearingDungeon()) {
+			int x = (width >> 1) - 64;
+			int y = (height >> 2) - 64;
+			hovered = DungeonMap.render(context, x, y, 1, true, mouseX - x, mouseY - y, hoveredElement(mouseX, mouseY).filter(PlayerButton.class::isInstance).map(PlayerButton.class::cast).map(p -> p.reference.uuid()).orElse(null));
+			context.drawBorder(x, y, 128, 128, -1);
+		}
 	}

```

3. And by the way, pressing E (inventory key) currently does not close the leap overlay. Would make sense to implement it.

kevinthegreat1 commented:
Note to self: one more bug, the tracking gets messed up when a player dyes. The solution should be to limit matches to the previous 100 or something.

kevinthegreat1 commented:
`Map<String, Table<Pair<UUID, String>, Long, MutableInt>>` to sort by timestamp as well?

kevinthegreat1 commented:
All bugs should have been fixed.

This could benefit from a dfu. This currently resets dungeons config.

AzureAaron commented:
> All bugs should have been fixed.
> 
> This could benefit from a dfu. This currently resets dungeons config.

There is no point, I doubt many people would ever be turning off the leap overlay to begin with.

kevinthegreat1 commented:
> There is no point, I doubt many people would ever be turning off the leap overlay to begin with.

The problem is this reset all dungeons config, unless we can think of a different name for the `spiritLeapOverlay` field.

layou233 commented on src/main/java/de/hysky/skyblocker/skyblock/dungeon/DungeonMap.java:
```diff
@@ -3,87 +3,226 @@
+				context.getMatrices().scale(2, 2, 1);
+				hovered = player.uuid();
+			}
+			RealmsUtil.drawPlayerHead(context, -4, -4, 8, player.uuid());
```
Is RealmsUtil good enough to use in game?
```java
public static void drawPlayerHead(DrawContext context, int x, int y, int size, UUID playerUuid) {
	MinecraftClient minecraftClient = MinecraftClient.getInstance();
	ProfileResult profileResult = minecraftClient.getSessionService().fetchProfile(playerUuid, false);
	SkinTextures skinTextures = profileResult != null
		? minecraftClient.getSkinProvider().getSkinTextures(profileResult.profile())
		: DefaultSkinHelper.getSkinTextures(playerUuid);
	PlayerSkinDrawer.draw(context, skinTextures, x, y, size);
}
```
fetchProfile(_, false) will load profile from cache, but would block the thread to fetch the profile via HTTP request if the cache misses.

I'm experiencing occasional game freezing with fancy dungeon map so I'm investigating thread blockers. 😢

kevinthegreat1 replied:
Yes, this does block the thread. I tried to pre-fetch skins on dungeon load, and I haven't seen the game freeze, but that's probably not the best solution.

We will need a few mixins to separate blocking fetches and non-blocking accesses, but it should be doable. You can also PR into this PR if you would like.

layou233 replied:
Or just copy this util function? And replace fetchProfile with SkullBlockEntity.fetchProfileByUuid().orElse(null), like what we've implemented in PartyEntry?

layou233 replied:
This works, I think. Played a few runs and haven't found any flaw yet.🥲
```diff
@@ -164,7 +170,11 @@
 				context.getMatrices().scale(2, 2, 1);
 				hovered = player.uuid();
 			}
-			RealmsUtil.drawPlayerHead(context, -4, -4, 8, player.uuid());
+			GameProfile playerProfile = SkullBlockEntity.fetchProfileByUuid(player.uuid).getNow(Optional.empty()).orElse(null);
+			SkinTextures skinTextures = playerProfile != null
+					? MinecraftClient.getInstance().getSkinProvider().getSkinTextures(playerProfile)
+					: DefaultSkinHelper.getSkinTextures(player.uuid);
+			PlayerSkinDrawer.draw(context, skinTextures, -4, -4, 8);
 			context.drawBorder(-5, -5, 10, 10, ColorHelper.fullAlpha(dungeonClass.color()));
```

kevinthegreat1 replied:
Done in cba3d56.

AzureAaron commented on src/main/java/de/hysky/skyblocker/utils/render/HudHelper.java:
```diff
@@ -0,0 +1,46 @@
+	 * This fetches the profile so it will be available for future calls to this method.
+	 */
+	public static void drawPlayerHead(DrawContext context, int x, int y, int size, String name) {
+		SkullBlockEntity.fetchProfileByName(name).getNow(Optional.empty()).map(CLIENT.getSkinProvider()::getSkinTextures).ifPresent(skinTexture -> PlayerSkinDrawer.draw(context, skinTexture, x, y, size));
```
This method had problems being used in the profile viewer where it would return an offline GameProfile (which also gets cached in the name cache for like a month), our method to fetch a name should be preferred to avoid this issue.

kevinthegreat1 commented on src/main/resources/assets/skyblocker/lang/en_us.json:
```diff
@@ -223,7 +225,10 @@
 
   "skyblocker.config.dungeons.spiritLeapOverlay": "Spirit Leap Overlay",
-  "skyblocker.config.dungeons.spiritLeapOverlay.@Tooltip": "Shows a custom overlay for the Spirit Leap screen that makes leaping to teammates easy and consistent.",
+  "skyblocker.config.dungeons.spiritLeapOverlay.enableLeapOverlay": "Enable Spirit Leap Overlay",
```
I think eventually I want to rename the field back to `spiritLeapOverlay` so I'll just keep the translation keys.

layou233 commented:
517fb25b363f42ca3853f5199683507e7ba3c375 hides all arrows on the map and breaks the intended purpose of "Show Self Head" option. 😢

Teyxos commented:
If not more than 1 puzzle room gets updated from tab at the same time just apply it to the one thats discovered this was something I was thinking if you wanna implement it when multiple rooms get discovered at the same time

Fluboxer commented:
I just tried to move my HUD midrun because of this:
<img width="260" height="267" alt="изображение" src="https://github.com/user-attachments/assets/9b5f698e-7d2c-40b2-b3d5-da7deaf7ce3e" />

Opened config, looked up HUD, clicked on edit and at moment I clicked 2nd tab it crashed
Crash log: [https://mclo.gs/DCsMFB0](https://mclo.gs/DCsMFB0)
[here's](https://mclo.gs/bOPTMo8) a part of log when it crashed (its a bit different from crash log)

kevinthegreat1 commented:
@Fluboxer were you playing with aaron? It can't find AzureAaron's uuid for some reason. I have fixed the crash.

kevinthegreat1 commented:
Seems to be tab hud stuff messing with the player list. I'll leave that for tab hud to fix.

Bloxigus commented on src/main/java/de/hysky/skyblocker/skyblock/dungeon/DungeonMap.java:
```diff
@@ -17,73 +22,174 @@
+			// Get the corresponding dungeon player for the map decoration.
+			DungeonPlayerManager.DungeonPlayer dungeonPlayer = null;
+			// If the map decoration is the self player, use the first player in this list. The self player is always the first player in the list.
+			if (mapDecoration.getValue().type().value().equals(MapDecorationTypes.FRAME.value())) {
```
this can just be `mapDecoration.getValue().type() == MapDecorationTypes.FRAME`
🥺🥺🥺

kevinthegreat1 commented:
> When nick is enabled, the avatar cannot be displayed properly.

I checked the code, and `DungeonPlayerManager` seems to never worked with nicks. I think we need some way of getting the current player's nick.

AzureAaron reviewed (approved):
I think it is fine
